### PR TITLE
Wirelesscache

### PR DESCRIPTION
**Pull request by jrddunbr**
I converted the WirelessUtils.java file to Kotlin, added a getChunk() method to Coordonate, as well as new function parts for the revamped wireless raytrace code.

### CODE REVIEW

**Comment by jrddunbr** [comment]
Thanks to @Grissess's effective rewrite of the math portions of the code I started (I'm not great at this kind of math), the virtualDistance calculation appears to work as intended now. It is mathematically working better than before - the new code handles signal traversal through blocks far better, including parts of blocks, which is currently somewhat broken - it could miss or count doubly for blocks previously, depending on the angle. Now, we use the distance that it travels through a block, which is more accurate over longer distances.

The only thing I would check at this point is whether the cache is working as we wanted. I think that it isn't quite what we wanted yet, so we may want to check on that, and make any necessary clarifications. Particularly, I'm looking at [line 40 and 42 in this code](https://github.com/jrddunbr/ElectricalAge/blob/d6ca96334162ffbd9b63520fd81a36317acfcc58/src/main/java/mods/eln/sixnode/wirelesssignal/WirelessUtils.kt#L40) - that may not be quite correct. Other edits may need to take place elsewhere for the loaded/unloaded chunk stuff - but that may be handled by L47 in the same file.

We did discuss making the hardness of the block factor into the equation, but Baughn commented that some of the hardness's are rather wacky (look at stone and planks for example), so that should wait for another time. It would, however, not be difficult to add that feature in, given the code here. Perhaps that and/or #812 could be done at the same time in a future pull request.

**Comment by cm0x4D** [comment]
For WAILA integration we used Google guava's cache (See [here](https://github.com/Electrical-Age/ElectricalAge/blob/3a8c989a6943af0bfc5bc254f138f4f6d5a10d3c/src/main/java/mods/eln/integration/waila/WailaCache.java)) and it worked very well. Might be of use here too...

File: src/main/java/mods/eln/misc/FloatLinear.kt [hunk 1]
```diff
@@ -0,0 +1,130 @@
+package mods.eln.misc
+
+data class Vec3f(var x: Float, var y: Float, var z: Float) {
```
Comment (Baughn) [inline src/main/java/mods/eln/misc/FloatLinear.kt:3]: I would really like some comments here. I mean, yes, it's a 3-element vector class with linear operations, but the other classes down below are nowhere near as obvious.

File: src/main/java/mods/eln/misc/FloatLinear.kt [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    companion object {
+        fun fromEndpoints(a: Vec3f, b: Vec3f) = Line3f(a, b.sub(a))
+    }
+
+    fun pointAt(u: Float) = origin.add(vector.mul(u))
+    fun midpoint() = pointAt(0.5f)
+    fun opposite() = pointAt(1.0f)
+    fun length() = vector.mag()
+
+    fun allIntersectionsAlongU(): List<Float> {
```
Comment (Baughn) [inline src/main/java/mods/eln/misc/FloatLinear.kt:61]: If you got this algorithm from somewhere, please include a link. That goes for all of this.

File: src/main/java/mods/eln/misc/FloatLinear.kt [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        val PLANE_Y = Plane3f(Vec3f.AXIS_Y, 0.0f)
+        val PLANE_Z = Plane3f(Vec3f.AXIS_Z, 0.0f)
+        val AXIS_PLANES = arrayOf(PLANE_X, PLANE_Y, PLANE_Z)
+    }
+
+    constructor(normal: Vec3f, point: Vec3f) : this(normal, -normal.dot(point))
+
+    fun pointOn() = normal.mul(-offset)
+
+    fun isctLineU(ln: Line3f): Float? {
```
Comment (Baughn) [inline src/main/java/mods/eln/misc/FloatLinear.kt:113]: What does "isctLine" do, exactly? Documentation is good, and better names would also be good.

File: src/main/java/mods/eln/sixnode/wirelesssignal/WirelessUtils.kt [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+            total += costMap.getOrDefault(chunk, 0.0)
+        }
+
+        cachedCost = total
+        return total
+    }
+}
+
+
+object WirelessUtils {
```
Comment (Baughn) [inline src/main/java/mods/eln/sixnode/wirelesssignal/WirelessUtils.kt:63]: This object probably shouldn't exist; it seems to have no state whatsoever. Functions can exist outside of classes, and can still be private or internal. I'm not saying it's the only way, but give it a bit of thought.

**Review by Baughn — CHANGES_REQUESTED** [review]

**Comment by jrddunbr** [comment]
I'm closing this because it's probably not going to happen for Minecraft 1.17.10. I'll keep the code locally on my GitHub until we're ready for it in the new Eln.

**Pull request closed without merging**